go.go

```diff
@@ -99,35 +99,55 @@ func Go(a, b []byte) (changes []*Change, err error) {
 			ds := x.diffstat(ai, bi)
 			changes = append(changes, &Change{Name: s, DelLines: ds.del, InsLines: ds.ins})
 		} else {
-			changes = append(changes, &Change{Name: s, Inserted: true, InsLines: len(bytes.Split(x.bBytes(bi), newline))})
+			changes = append(changes, &Change{Name: s, Inserted: true, InsLines: x.bLines(bi)})
 		}
 	}
 	for s, ai := range del {
 		if _, ok := ins[s]; !ok {
-			changes = append(changes, &Change{Name: s, Deleted: true, DelLines: len(bytes.Split(x.aBytes(ai), newline))})
+			changes = append(changes, &Change{Name: s, Deleted: true, DelLines: x.aLines(ai)})
 		}
 	}
 
 	if other != nil {
-		// Calculate an overall diffstat.
-		aLines := bytes.Split(a, newline)
-		bLines := bytes.Split(b, newline)
-		ab := diff.Bytes(aLines, bLines)
-		es := diff.Myers(context.Background(), ab)
-		ins, del := es.Stat()
-		// Subtract all diffs accountable for by other changes.
-		for _, c := range changes {
-			ins -= c.InsLines
-			del -= c.DelLines
+		// Calculate diffstats between non-declarations.
+		aIsLarger := false
+		lens := len(x.asplit) - 1
+		if len(x.asplit) > len(x.bsplit) {
+			lens = len(x.bsplit) - 1
+			aIsLarger = true
 		}
-		if ins < 0 {
-			ins = 0
+		// If df is increased, it means that an index has been displaced by some change.
+		var df int64
+		for i := 1; i < lens; i++ {
+			if int64(x.asplit[i]) != (int64(x.bsplit[i]) - df) {
+				df = abs(int64(x.asplit[i] - x.bsplit[i]))
```

Comment:
This seems like a poor-man's diff: You're trying to correlate left and right elements. But we've already done a diff. I suspect that you should be working off the edit script somehow, not off the raw data.

Also, I'm not sure how important it is to have the sum of the semdiffstat additions and deletions add up to the non-semantic diffstat numbers. Particularly with order-insensitive matches, they won't, because we can detect code movement and regular diff stat can't.

So let's focus on picking an algorithm that gives the user decent sense for how much else has changed. The goal here is semantic, after all, not syntactic.

As a notable example, if we have order-insensitive semantic diffstats, and you move a function and its associated comments, then that should contribute 0 to the "other" diffstat, but with the current setup, the comment movements will be viewed as generating a diff.

It is possible that we should try to use a go/ast comment map to associate comments with decls and then diff from there. (go/ast's handling of comments is kinda awful, but it is probably better than nothing.)

Reply:
> But we've already done a diff. I suspect that you should be working off the edit script

Good point, I'll see if I can use the edit script instead.

> the comment movements will be viewed as generating a diff.

I'll think about this.


> It is possible that we should try to use a go/ast comment map

I'm not familiar with go/ast, I'll look into it.